crates/wasm-mutate/src/mutators/peephole/eggsy/mod.rs

```diff
@@ -0,0 +1,284 @@
+use std::{cmp::Ordering, collections::HashMap};
+
+use egg::{define_language, Analysis, CostFunction, EClass, EGraph, Id, Language, Symbol};
+use rand::Rng;
+
+pub mod analysis;
+pub mod lang;
+
+use crate::mutators::peephole::eggsy::lang::Lang;
+
+pub struct NoPopcnt;
+
+impl CostFunction<Lang> for NoPopcnt {
+    type Cost = usize;
+    fn cost<C>(&mut self, enode: &Lang, mut costs: C) -> Self::Cost
+    where
+        C: FnMut(Id) -> Self::Cost,
+    {
+        let op_cost = match enode {
+            // Doubts on this...
+            Lang::I32Popcnt(_) => usize::MAX,
+            _ => 0,
+        };
+        enode.fold(op_cost, |sum, id| sum.saturating_add(costs(id)))
+    }
+}
+
+/// This struct is a wrapper of egg::Extractor
+/// The majority of the methods are copied and adapted to our needs
+pub struct RandomExtractor<'a, CF: CostFunction<L>, L: Language, N: Analysis<L>> {
+    cost_function: CF,
+    costs: HashMap<Id, (CF::Cost, usize)>,
+    egraph: &'a EGraph<L, N>,
+}
+
+fn cmp<T: PartialOrd>(a: &Option<T>, b: &Option<T>) -> Ordering {
+    // None is high
+    match (a, b) {
+        (None, None) => Ordering::Equal,
+        (None, Some(_)) => Ordering::Greater,
+        (Some(_), None) => Ordering::Less,
+        (Some(a), Some(b)) => a.partial_cmp(&b).unwrap(),
+    }
+}
+
+impl<'a, CF, L, N> RandomExtractor<'a, CF, L, N>
+where
+    CF: CostFunction<L>,
+    L: Language,
+    N: Analysis<L, Data = Option<i32>>, // The analysis should return the index of the node in the e-class
+{
+    pub fn new(egraph: &'a EGraph<L, N>, cost_function: CF) -> Self {
+        let costs = HashMap::default();
+
+        let mut extractor = RandomExtractor {
+            costs,
+            egraph,
+            cost_function,
+        };
+        extractor.costs = extractor.find_costs();
+        extractor
+    }
+
+    fn find_costs(&mut self) -> HashMap<Id, (CF::Cost, usize)> {
```

Comment:
Now here, on the other hand, we are keeping track of the node's index, which is what the `usize` is here.